tests/v2/test_0835-datetime-type.py

```diff
@@ -47,454 +53,692 @@ def test_date_time_minmax():
+)
+def test_date_time_units():
+    array1 = np.array(
+        ["2020-07-27T10:41:11", "2019-01-01", "2020-01-01"], "datetime64[s]"
+    )
+    array2 = np.array(
+        ["2020-07-27T10:41:11", "2019-01-01", "2020-01-01"], "datetime64[25s]"
+    )
+    ak_a1 = ak.Array(array1)
+    ak_a1 = v1_to_v2(ak_a1.layout)
+
+    ak_a2 = ak.Array(array2)
+    ak_a2 = v1_to_v2(ak_a2.layout)
+
+    np_ar1 = ak._v2.operations.convert.to_numpy(ak_a1)
+    np_ar2 = ak._v2.operations.convert.to_numpy(ak_a2)
+
+    if np_ar1[0] > np_ar2[0]:
+        assert (np_ar1[0] - np.timedelta64(25, "s")) < np_ar2[0]
+    else:
+        assert (np_ar1[0] + np.timedelta64(25, "s")) >= np_ar2[0]
+
+
+def test_sum():
+
+    dtypes = ["datetime64[s]", "timedelta64[D]"]
+
+    arrays = (np.arange(0, 12, dtype=dtype) for dtype in dtypes)
+    for array in arrays:
+        content = ak._v2.contents.NumpyArray(array)
+        offsets = ak._v2.index.Index64(np.array([0, 4, 8, 12], dtype=np.int64))
+        depth = ak._v2.contents.ListOffsetArray(offsets, content)
+
+        if np.issubdtype(array.dtype, np.timedelta64):
+            assert ak.to_list(depth.sum(-1)) == [
+                datetime.timedelta(6),
+                datetime.timedelta(22),
+                datetime.timedelta(38),
+            ]
+
+            assert ak.to_list(depth.sum(1)) == [
+                datetime.timedelta(6),
+                datetime.timedelta(22),
+                datetime.timedelta(38),
+            ]
+
+            assert ak.to_list(depth.sum(-2)) == [
+                datetime.timedelta(12),
+                datetime.timedelta(15),
+                datetime.timedelta(18),
+                datetime.timedelta(21),
+            ]
+            assert ak.to_list(depth.sum(0)) == [
+                datetime.timedelta(12),
+                datetime.timedelta(15),
+                datetime.timedelta(18),
+                datetime.timedelta(21),
+            ]
+
+        else:
+            with pytest.raises(ValueError):
+                depth.sum(-1)
+
+
+def test_more():
+    nparray = np.array(
+        [np.datetime64("2021-06-03T10:00"), np.datetime64("2021-06-03T11:00")]
+    )
+    akarray = ak._v2.highlevel.Array(nparray)
+
+    assert (akarray[1:] - akarray[:-1]).tolist() == [np.timedelta64(60, "m")]
+    assert ak.sum(akarray[1:] - akarray[:-1]) == np.timedelta64(60, "m")
+    assert ak.sum(akarray[1:] - akarray[:-1], axis=0) == [np.timedelta64(60, "m")]
+
+
+def test_ufunc_sum():
+    nparray = np.array(
+        [np.datetime64("2021-06-03T10:00"), np.datetime64("2021-06-03T11:00")]
+    )
+    akarray = ak._v2.highlevel.Array(nparray)
+
+    with pytest.raises(np.core._exceptions.UFuncTypeError):
+        akarray[1:] + akarray[:-1]
+
+
+def test_ufunc_mul():
+    nparray = np.array(
+        [np.datetime64("2021-06-03T10:00"), np.datetime64("2021-06-03T11:00")]
+    )
+    akarray = ak._v2.highlevel.Array(nparray)
+
+    with pytest.raises(np_exception._UFuncBinaryResolutionError):
```

Comment:
this does not work with earlier versions of Numpy. In any case, it should be a `ValueError` exception.

Reply:
Same comment as below. It's because `np.core._exceptions` is private. This one descends from TypeError, not ValueError, though:

```python
>>> np.core._exceptions._UFuncBinaryResolutionError.__mro__
(<class 'numpy.core._exceptions._UFuncBinaryResolutionError'>,
 <class 'numpy.core._exceptions.UFuncTypeError'>,
 <class 'TypeError'>,
 <class 'Exception'>,
 <class 'BaseException'>,
 <class 'object'>)
```

Reply:
done